### PR TITLE
GMRFfskyride interval refactor

### PR DESCRIPTION
**Pull request by jtmccr1**
TL;DR
This abstracts trees out of skyride and skygrid likelihoods so that we can get rid of code duplication and use alternative implementations of interval lists.

## Aim
 We would like to add a layer of abstraction so that coalescent likelihoods do not need to  “know” about a tree, instead they would only know about coalescent intervals.  This pull request does just  that for GMRFSkyride and GMRFSkygrid likelihoods.  Doing this has affected classes that I am not very familiar with (mostly in the gradient/HMC realm), and I’d like to make sure everything is OK with those developing those classes before merging. If this is very disruptive it may be best to keep this as a separate branch for the time being, or think about a different solution.

## Motivation 
There have been a handful of cases where it was useful to have alternative implementations of coalescent intervals. But using these implementations with a skygrid coalescent likelihood has not been possible because `GMRFMultilocusSkyrideLikelihood` and `GMRFSkyrideLikelihood`  extend  `OldAbstractCoalescentLikelihood` which wraps a Tree and is a  coalescent interval provider.  To use these implementations we’ve had to make parallel classes which extend `AbstractCoalescentLikelihood` and take `IntervalLists` instead of trees, but obviously this duplication is not ideal. It also means that it has been impossible to use these coalescent intervals with HMC operators, which rely on the interval-node mapping provided by `OldAbstractCoalescentLikelihood`. 

_NB: There are two `TreeIntervals` classes one in evolution.coalescent and one in evomodel.coalescent. The Skygrid  used to use the one in evolution.coalescent. This class starts with a “nothing” interval that ends with the first sample event; whereas, the model class, (and all others as far as I can tell) starts at the first sample event. I have refactored the skygrid to use the model and adjusted the indexing in `setupSufficientStatistics`
 accordingly._

## Proposed Changes
  -Remove skyride and skygrid’s dependance on `OldAbstractCoalescentLikelihood` and instead extend `AbstractCoalescentLikelihood`.
  -Replace the trees in skyride and skygrid with intervalLists 
  -extract interval-node mapping functionality to a new interface `TreeIntervalList` which wraps a tree and is itself is an `IntervalList` .
-Update the gradient classes which use the interval-node mapping to use this new interface.
	-Update these parsers to verify that the intervalList in the sky*likelihood can be cast into a TreeIntervalList (i.e. verify the interval-node mapping is available during parsing) .
-Upate TreeIntervals (in evomodel.coalescent) to implement `TreeIntervalList`
-Upate  BigFastTreeIntervals to implement  `TreeIntervalList`
-Remove the now-redundant, duplicated classes `GMRFSkygridLikelihood` 

In the future remove:  `OldGMRFSkyrideLikelihood`

## Does it work?
It seems to.

I have simple unit tests that check the `GMRFSkyride`, `GMRFMultilocusSkygrid`, and `GMRFSkyrideGradient` return the same loglikelihoods in this branch compared to the master.

I have a unit test that verifies TreeIntervals and BigFastTreeIntervals properly implement the interval-node mappings. 

I have added a testSkygrid.xml that verifies the updated `GMRFmulitlocusSkygrid` returns the same loglikelihood as the BEASTv1.10.4 release.

I have added `testGMRFBayesianSkygrid+strict.xml` and `testGMRFBayesianSkyride+strict.xml` to the ci test directory (adapted from examples/testXML) and these give the same log likelihoods as the current master branch after a short run.  As does `testGMRFBayesianSkyride+HMC+strict.xml` which includes the skygrid HMC operator.

---
Something I am not sure about: The `TreeIntervals` in `evolution.coalsescent` has a multifurcation limit which I believe allows for merging very small sampling intervals. This behavior is not replicated in this pull request. It is possible to have intervals of duration 0. I’m not sure if that will lead to down-stream numerical issues.

I would appreciate any feedback. I don’t want to disturb other’s work. Do these changes make sense to you? If anyone has any tests or example xmls, I’d be happy to make them into a test case to verify nothing has gone awry.

### CODE REVIEW

**Comment by rambaut** [comment]
Could we call the new classes just `SkyrideLikelihood` and `SkygridLikelihood`? This is what everyone knows them as. The parser could have a synonym for backwards compatibility?

**Comment by jtmccr1** [comment]
I agree with Andrew. I think it might make even more sense if the GMRFs ever were refactored out of those classes, but that may be discussion for another time (and probably not for me :) ) 

I'm starting to think master might not be the best place for this yet. If no one opposes the concept, I could move it to the bigfast trees branch and then coordinate merging that branch with one of the hmc branches. That would be a more direct/natural way to unify the work Xiang and I have been doing. It would also be more gradual and give us more chances to catch any potential problems.

**Comment by GuyBaele** [comment]
Agree to put this in a branch for the time being, given the link with the work on HMC. It would be good to have these changes, so as to remove dependencies on OldAbstractCoalesentLikelihood, and indeed move to having one TreeIntervals class. Not sure about the multifurcation limit; Andrew and Marc may know about the need for (not) keeping this.

**Comment by jtmccr1** [comment]
After chatting with Guy and Xiang, I think it is best to keep this work as a separate branch for the time being. We can think about merging it in later.
